### PR TITLE
[apm] add peer tag aggregation to trace agent

### PR DESCRIPTION
**Pull request by jdgumz**
### What does this PR do?

Draws heavy inspiration from https://github.com/DataDog/datadog-agent/pull/17933/files and adds a field that allows for new peer tags to be included in APM stats aggregation.

### Motivation

Allow for expansion of the trace metric data model to assist with better describing `peer.service` entities. Note: the backend will still make decisions on which tags to accept and place on trace metrics. Arbitrary peer tags are not supported.

### Possible Drawbacks / Trade-offs

Additional aggregation overhead in the APM stats codepath.

### Describe how to test/QA your changes

Submit data from a tracer that includes spans which:
- have `span.kind: client` or `span.kind: producer`
- have one or more of the fields in the span `meta` that you intend to set in `apm_config.peer_tags`
- enable `apm_config.peer_service_aggregation`

Example of new peer tags getting added (e.g., `aws.s3.bucket`):
image

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/config/config.go [hunk 1]
```diff
@@ -502,6 +503,9 @@ func New() *AgentConfig {
 
 		BucketInterval: time.Duration(10) * time.Second,
 
+		// TODO: Separate additional "defaults" from custom tags once aggregation keys are fully dynamic.
+		CustomTags: []string{"aws.s3.bucket"},
```
Comment (jdgumz) [inline pkg/trace/config/config.go:507]: nit: move any "defaults" to where we initialize the config

File: comp/trace/config/setup.go [hunk 2]
```diff
@@ -223,6 +224,22 @@ func applyDatadogConfig(c *config.AgentConfig, core corecompcfg.Component) error
 	}
 	c.PeerServiceAggregation = core.GetBool("apm_config.peer_service_aggregation")
 	c.ComputeStatsBySpanKind = core.GetBool("apm_config.compute_stats_by_span_kind")
+	if core.IsSet("apm_config.custom_tags") {
+		// TODO: make func, test it separately
```
Comment (jdgumz) [inline comp/trace/config/setup.go:228]: TODO: add defaults here

If the field is set, then we can consider that to be opt-in.

File: pkg/trace/stats/concentrator.go [hunk 3]
```diff
@@ -48,6 +49,27 @@ type Concentrator struct {
 	agentVersion           string
 	peerSvcAggregation     bool // flag to enable peer.service aggregation
 	computeStatsBySpanKind bool // flag to enable computation of stats through checking the span.kind field
+	extraTags              []string
+}
+
+// TODO: finish full list
+// PeerServiceEntityTags represents the full list of supplementary tags to describe a peer.service entity.
+var PeerServiceEntityTags = []string{"rpc.host", "grpc.host", "db.instance", "db.system", "kafka.topic"}
```
Comment (jdgumz) [inline pkg/trace/stats/concentrator.go:57]: @djova I worry a little bit about the cost if this slice grows to become 20+ elements. We might need to get smarter over time, like an index of `op_name: [entity_tags]`.

File: pkg/proto/datadog/trace/stats.proto [hunk 4]
```diff
@@ -72,4 +72,5 @@ message ClientGroupedStats {
 	uint64 topLevelHits = 13; // count of top level spans aggregated in the groupedstats
 	string peer_service = 14; // name of the remote service that the `service` communicated with
 	string span_kind = 15; // value of the span.kind tag on the span
+	repeated string extra_tags = 16; // specifies extra tags to be used for stats aggregation
```
Comment (pkalmakis) [inline pkg/proto/datadog/trace/stats.proto:75]: Perhaps we can just call this `tags`?

File: pkg/proto/pbgo/core/api.pb.go [hunk 5]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.31.0
-// 	protoc        v3.21.5
+// 	protoc-gen-go v1.28.1
```
Comment (pkalmakis) [inline pkg/proto/pbgo/core/api.pb.go:3]: It's probably safest to exclude these unrelated files from your change list.

File: pkg/trace/config/config.go [hunk 6]
```diff
@@ -338,6 +338,7 @@ type AgentConfig struct {
 	ExtraAggregators       []string      // DEPRECATED
 	PeerServiceAggregation bool          // enables/disables stats aggregation for peer.service, used by Concentrator and ClientStatsAggregator
 	ComputeStatsBySpanKind bool          // enables/disables the computing of stats based on a span's `span.kind` field
+	CustomTags             []string      // [EXPERIMENTAL] additional tags to use for stats aggregation
```
Comment (pkalmakis) [inline pkg/trace/config/config.go:341]: Should we reuse the `ExtraAggregators` field? Perhaps its unsafe to do? I'm worried about calling these `Custom` tags because we are not yet ready to support things that are "custom".

File: pkg/trace/stats/aggregation.go [hunk 7]
```diff
@@ -66,7 +68,7 @@ func getStatusCode(s *pb.Span) uint32 {
 }
 
 // NewAggregationFromSpan creates a new aggregation from the provided span and env
-func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregationKey, enablePeerSvcAgg bool) Aggregation {
+func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregationKey, enablePeerSvcAgg bool, extraTags []string) Aggregation {
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:71]:
```suggestion
func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregationKey, enablePeerSvcAgg bool, extraTagKeys []string) Aggregation {
```

File: pkg/trace/stats/aggregation.go [hunk 8]
```diff
@@ -25,6 +25,8 @@ const (
 type Aggregation struct {
 	BucketsAggregationKey
 	PayloadAggregationKey
+	// ExtraTagsKey is a comma-delimited string of all tag:value pairs for additional tags that should be used in stats aggregation.
+	ExtraTagsKey string
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:29]: Using a delimited string can break very easy if the delimiter is present in any of the tag values (which is entirely out of our control). Additionally, doing string manipulations for every aggregation could bring overhead of a bunch of extra allocations. Perhaps we should think about keying off a hash of the sorted tags?

File: pkg/trace/stats/aggregation.go [hunk 9]
```diff
@@ -83,12 +85,30 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
 	}
+	agg.ExtraTagsKey = extraTagsKey(s, extraTags)
 	return agg
 }
 
+func extraTagsKey(s *pb.Span, extraTags []string) string {
+	if len(extraTags) == 0 {
+		return ""
+	}
+	// TODO: replace with strings.Builder?
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:96]: If we stick this this approach, we indeed probably want to use a `strings.Builder`, sort the tags, and choose a less-common character as a delimiter. (And potentially escape delimiters that are present in the tags)

File: pkg/trace/stats/aggregation.go [hunk 10]
```diff
@@ -99,4 +119,9 @@ func NewAggregationFromGroup(g *pb.ClientGroupedStats) Aggregation {
 			Synthetics:  g.Synthetics,
 		},
 	}
+
+	if g.ExtraTags != nil {
+		agg.ExtraTagsKey = strings.Join(g.ExtraTags, ",")
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:124]: These would have to be sorted.

File: pkg/trace/stats/concentrator.go (hunk 3 above)
Comment (pkalmakis) [reply pkg/trace/stats/concentrator.go:57]: Let's try to be as smart as possible from the jump so that we don't have to check these additional tags for every single span we see.

**Review by pkalmakis — CHANGES_REQUESTED** [review]

File: pkg/trace/config/config.go (hunk 6 above)
Comment (jdgumz) [reply pkg/trace/config/config.go:341]: I wondered about this but then thought it could be unsafe for customers who may have old `ExtraAggregators` config lying around?

File: pkg/trace/stats/aggregation.go (hunk 8 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:29]: You're right, although we used to use such delimiters and strings in the old aggregation map long ago.

We could add the hash here and then add the `[]string` of custom tags to the `groupedStats` struct.

File: pkg/trace/stats/aggregation.go (hunk 10 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:124]: True, this assumes that the client sent a sorted tag string.

File: pkg/trace/stats/aggregation.go (hunk 9 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:96]: :+1: The sorting is handled already during the initialization of the concentrator.

File: pkg/trace/stats/aggregation.go [hunk 11]
```diff
@@ -65,8 +70,27 @@ func getStatusCode(s *pb.Span) uint32 {
 	return uint32(c)
 }
 
+// stringToBytes unsafely casts a string to a byte slice.
+func stringToBytes(str string) []byte {
```
Comment (jdgumz) [inline pkg/trace/stats/aggregation.go:74]: This should be fairly safe because:
1. We're not modifying the returned `[]byte`.
2. The only place the `[]byte` is used (read) is for the hashing function.

File: pkg/trace/stats/statsraw_test.go [hunk 12]
```diff
@@ -84,15 +86,110 @@ func TestGrainWithPeerService(t *testing.T) {
...
+				Hostname:    "default",
+				ContainerID: "cid",
+			},
+			BucketsAggregationKey: BucketsAggregationKey{
+				Service:      "thing",
+				SpanKind:     "client",
+				Name:         "other",
+				Resource:     "yo",
+				PeerService:  "aws-s3",
+				PeerTagsHash: 6956349601612342508,
```
Comment (jdgumz) [inline pkg/trace/stats/statsraw_test.go:146]: Not a huge fan of hard-coding the values. I could make a func that does the fnv hash but that's pretty close already to the underlying func.

File: pkg/proto/pbgo/core/api.pb.go (hunk 5 above)
Comment (jdgumz) [reply pkg/proto/pbgo/core/api.pb.go]: @truthbk explained that this should be fine - will tag him for review.

**Review by ogaca-dd — APPROVED** [review]

File: pkg/proto/pbgo/core/api.pb.go [hunk 13]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.31.0
-// 	protoc        v3.21.5
+// 	protoc-gen-go v1.30.0
```
Comment (truthbk) [inline pkg/proto/pbgo/core/api.pb.go:3]: This is interesting.... Not sure why the compiler version is lower. The `inv -e install-tools` was supposed to have kept us aligned; there might be a bug or something in that invoke task. 

Can you confirm you called `inv -e install-tools` prior to generating the new definitions? If you did, then there's nothing for you to do, I can take a look and submit a follow-up PR if necessary.

**Review by truthbk — APPROVED** [review]
I think this looks good on the protobuf front. Small question with regard to the `protoc-gen-go` version, but likely just a NOP for you. 👍

File: pkg/proto/pbgo/core/api.pb.go (hunk 13 above)
Comment (jdgumz) [reply pkg/proto/pbgo/core/api.pb.go:3]: Yes I did call this at least once before regenerating the protobuf. That is curious...

**Review by alai97 — APPROVED** [review]
Looks good for docs!

File: pkg/proto/pbgo/mocks/core/api_mockgen.pb.go [hunk 14]
```diff
@@ -1,5 +1,5 @@
 // Code generated by MockGen. DO NOT EDIT.
-// Source: /workspaces/datadog-agent/pkg/proto/pbgo/core/api.pb.go
+// Source: /Users/jesse.gumz/go/src/github.com/DataDog/datadog-agent/pkg/proto/pbgo/core/api.pb.go
```
Comment (pkalmakis) [inline pkg/proto/pbgo/mocks/core/api_mockgen.pb.go:2]: Might want to revert this?

File: pkg/trace/stats/aggregation.go [hunk 15]
```diff
@@ -29,14 +32,15 @@ type Aggregation struct {
...
-	Name        string
-	PeerService string
-	Resource    string
-	Type        string
-	SpanKind    string
-	StatusCode  uint32
-	Synthetics  bool
+	Service      string
+	Name         string
+	PeerService  string
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:37]: Are we wanting this to still be a separate tag?

File: pkg/trace/stats/aggregation.go [hunk 16]
```diff
@@ -65,8 +69,20 @@ func getStatusCode(s *pb.Span) uint32 {
...
+func stringToBytes(s string) []byte {
+	return unsafe.Slice(unsafe.StringData(s), len(s))
+}
+
+func spanKindIsClientOrProducer(spanKind string) bool {
+	sk := strings.ToLower(spanKind)
+	if sk == "client" || sk == "producer" {
+		return true
+	}
+	return false
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:81]: `return sk == "client" || sk == "producer"`

File: pkg/trace/stats/aggregation.go [hunk 17]
```diff
@@ -80,15 +96,48 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
 	}
-	return agg
+	return agg, peerTags
+}
+
+func getMatchingPeerTags(s *pb.Span, peerTagKeys []string) []string {
+	if len(peerTagKeys) == 0 {
+		return nil
+	}
+	pt := make([]string, 0)
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:114]: `var pt []string`

File: pkg/trace/stats/aggregation.go [hunk 18]
```diff
@@ -80,15 +96,48 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+	return pt
+}
+
+func peerTagsHash(tags []string) uint64 {
+	if len(tags) == 0 {
+		return 0
+	}
+	h := fnv.New64a()
+	for _, t := range tags {
+		h.Write(stringToBytes(t))
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:133]: I believe the go compiler knows to allocate a byte array in the stack in this case because it isn't reused and you can just do `[]byte(t)`. Not sure it makes a difference either way but it might save having to reimplement and use this unsafe function.

File: pkg/trace/stats/aggregation.go [hunk 19]
```diff
@@ -99,4 +148,12 @@ func NewAggregationFromGroup(g *pb.ClientGroupedStats) Aggregation {
 			Synthetics:  g.Synthetics,
 		},
 	}
+
+	if g.PeerTags != nil {
+		if !sort.StringsAreSorted(g.PeerTags) {
+			sort.Strings(g.PeerTags)
+		}
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:155]: Move this to inside peerTagsHash in case there are other callers in the future?

File: pkg/trace/stats/aggregation.go (hunk 18 above)
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:133]: You still want to use a delimiter here, just not a comma. It should be safe to just write `[]byte{0}` as the 0 byte should never be in a string

File: pkg/proto/pbgo/mocks/core/api_mockgen.pb.go (hunk 14 above)
Comment (jdgumz) [reply pkg/proto/pbgo/mocks/core/api_mockgen.pb.go:2]: Thanks cc @truthbk it seems like there might be something we want to fix in terms of how the scripting is working here.

File: pkg/trace/stats/aggregation.go (hunk 15 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:37]: It's a fair question. I don't think we have to put it with the rest of `PeerTags` today but that could be a future change.

File: pkg/trace/stats/aggregation.go (hunk 19 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:155]: Good idea!

Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:155]: Although this may incur some more CPU cycles, it might be better to just make things more "programmer safe".

File: pkg/trace/stats/aggregation.go (hunk 15 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:36]: Also: the `peer_tags` field is meant to be an extra config opt-in. It would be weird to have it be "on" anyway because of `peer.service`?

File: pkg/config/apm.go [hunk 20]
```diff
@@ -164,6 +164,15 @@ func setupAPM(config Config) {
 		}
 		return out
 	})
+
+	config.BindEnv("apm_config.peer_tags", "DD_APM_PEER_TAGS")
```
Comment (katiehockman) [inline pkg/config/apm.go:168]: Consider adding this to `pkg/flare/envvars.go` if you want it included in an Agent Flare (which I think you probably do?)

File: pkg/config/config_template.yaml [hunk 21]
```diff
@@ -1480,6 +1480,13 @@ api_key:
   ## may not be marked by the Agent as top-level spans.
   # peer_service_aggregation: false
 
+  ## @param peer_tags - list of strings - optional
+  ## @env DD_APM_PEER_TAGS - list of strings - optional
+  ## Enables additional tags related to peer.service. Will not be used unless peer_service_aggregation is also set.
```
Comment (katiehockman) [inline pkg/config/config_template.yaml:1485]: > Enables additional tags related to peer.service.

Do you think customers will have enough context to know what this means? I personally doubt that most would. Is there a link to some public docs that describe this that you can link here?

File: pkg/config/config_template.yaml [hunk 22]
```diff
@@ -1480,6 +1480,13 @@ api_key:
...
+  ## NOTE: This option is recommended to get the most granular tagging alongside `peer.service`, but it will
+  ## also increase the computational overhead of aggregation. This list can be omitted if that cost is too high for your agent.
+  # peer_tags: false
```
Comment (katiehockman) [inline pkg/config/config_template.yaml:1488]: > peer_tags: false

This shouldn't be a boolean here, right? Isn't it a list of strings?

File: pkg/proto/datadog/trace/stats.proto [hunk 23]
```diff
@@ -72,4 +72,5 @@ message ClientGroupedStats {
 	uint64 topLevelHits = 13; // count of top level spans aggregated in the groupedstats
 	string peer_service = 14; // name of the remote service that the `service` communicated with
 	string span_kind = 15; // value of the span.kind tag on the span
+	repeated string peer_tags = 16; // supplementary tags that further describe peer_service
```
Comment (katiehockman) [inline pkg/proto/datadog/trace/stats.proto:75]: > // supplementary tags that further describe peer_service

I don't feel this is descriptive enough. What does "further describe peer_service" mean?

File: pkg/trace/stats/aggregation.go [hunk 24]
```diff
@@ -65,8 +68,13 @@ func getStatusCode(s *pb.Span) uint32 {
 	return uint32(c)
 }
 
+func spanKindIsClientOrProducer(spanKind string) bool {
```
Comment (katiehockman) [inline pkg/trace/stats/aggregation.go:71]:
```suggestion
func clientOrProducer(spanKind string) bool {
```

nit: just a personal preference to shorten this a bit

File: pkg/trace/stats/aggregation.go [hunk 25]
```diff
@@ -80,15 +88,54 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+			agg.PeerService = s.Meta[tagPeerService]
+		}
+		peerTags = getMatchingPeerTags(s, peerTagKeys)
+		agg.PeerTagsHash = peerTagsHash(peerTags)
 	}
-	return agg
+	return agg, peerTags
+}
+
+func getMatchingPeerTags(s *pb.Span, peerTagKeys []string) []string {
```
Comment (katiehockman) [inline pkg/trace/stats/aggregation.go:102]: https://go.dev/doc/effective_go#Getters

should just be `matchingPeerTags`

File: pkg/trace/stats/aggregation.go [hunk 26]
```diff
@@ -80,15 +88,54 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+	var pt []string
+	for _, t := range peerTagKeys {
+		v, ok := s.Meta[t]
+		if ok {
+			pt = append(pt, t+":"+v)
+		}
+	}
+	if len(pt) == 0 {
+		return nil
+	}
```
Comment (katiehockman) [inline pkg/trace/stats/aggregation.go:115]: No need for this. If the loop never matches anything, then `pt` will be nil anyway

File: pkg/trace/stats/aggregation.go [hunk 27]
```diff
@@ -80,15 +88,54 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+func getMatchingPeerTags(s *pb.Span, peerTagKeys []string) []string {
+	if len(peerTagKeys) == 0 {
+		return nil
+	}
+	var pt []string
+	for _, t := range peerTagKeys {
+		v, ok := s.Meta[t]
+		if ok {
+			pt = append(pt, t+":"+v)
+		}
```
Comment (katiehockman) [inline pkg/trace/stats/aggregation.go:111]:
```suggestion
		if v, ok := s.Meta[t]; ok {
			pt = append(pt, t+":"+v)
		}
```

File: pkg/trace/stats/aggregation.go [hunk 28]
```diff
@@ -80,15 +88,54 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+	if !sort.StringsAreSorted(tags) {
+		sort.Strings(tags)
+	}
+	h := fnv.New64a()
+	for i, t := range tags {
+		h.Write([]byte(t))
+		if i != len(tags)-1 {
+			h.Write([]byte{0})
+		}
+	}
```
Comment (katiehockman) [inline pkg/trace/stats/aggregation.go:132]: Could you just use `strings.Join` and then do the hash of that, instead? Might simplify this a bit
(https://pkg.go.dev/strings#Join)

File: pkg/trace/stats/aggregation.go [hunk 29]
```diff
@@ -99,4 +146,9 @@ func NewAggregationFromGroup(g *pb.ClientGroupedStats) Aggregation {
 			Synthetics:  g.Synthetics,
 		},
 	}
+
+	if g.PeerTags != nil {
```
Comment (katiehockman) [inline pkg/trace/stats/aggregation.go:150]: I don't think you need this `nil` check. If `g.PeerTags` is in fact nil, then `peerTagsHash` will return 0 since you do a length check at the beginning of the function. And since the zero-value for `agg.PeerTagsHash` is in fact `0`, then it won't change the value if you set `agg.PeerTagsHash = peerTagsHash(g.PeerTags)` even if `peerTagsHash` returns `0`.

So basically, I'm suggesting just setting this the same way as the rest above.

File: pkg/trace/stats/aggregation_test.go [hunk 30]
```diff
@@ -57,33 +57,57 @@ func TestNewAggregation(t *testing.T) {
...
 			},
 			false,
+			Aggregation{BucketsAggregationKey: BucketsAggregationKey{Service: "a", SpanKind: "client"}},
+			nil,
+		},
+		// peer.service stats aggregation is enabled, but span.kind != (client, producer).
+		{
+			&pb.Span{
+				Service: "a",
+				Meta:    map[string]string{"span.kind": "", "peer.service": "remote-service"},
```
Comment (katiehockman) [inline pkg/trace/stats/aggregation_test.go:88]: I suggest one where `span.kind` is empty (as you have) but another where `span.kind` is non-empty but set to something other than `client` or `producer`

File: pkg/trace/stats/aggregation.go (hunk 29 above)
Comment (katiehockman) [reply pkg/trace/stats/aggregation.go:150]: Unless you need to check `enablePeerSvcAgg` before setting this value (which isn't happening right now, but I see in other parts of the code?)

File: pkg/trace/stats/concentrator.go [hunk 31]
```diff
@@ -46,8 +47,25 @@ type Concentrator struct {
...
+	}
+	ptm := make(map[string]struct{})
+	for _, t := range tags {
+		ptm[t] = struct{}{}
+	}
+	deduped := make([]string, 0, len(ptm))
+	for t := range ptm {
+		deduped = append(deduped, t)
+	}
+	sort.Strings(deduped)
```
Comment (katiehockman) [inline pkg/trace/stats/concentrator.go:67]: If you're sorting them here, then do you need to sort it in the other function earlier in this PR?

File: pkg/trace/stats/concentrator.go [hunk 32]
```diff
@@ -46,8 +47,25 @@ type Concentrator struct {
...
+	peerSvcAggregation     bool     // flag to enable peer.service aggregation
+	computeStatsBySpanKind bool     // flag to enable computation of stats through checking the span.kind field
+	peerTagKeys            []string // keys for supplementary tags that describe peer.service entities
+}
+
+func prepareTagKeys(tags ...string) []string {
+	if len(tags) == 0 {
+		return nil
+	}
+	ptm := make(map[string]struct{})
```
Comment (katiehockman) [inline pkg/trace/stats/concentrator.go:59]: Why a `struct{}`? A `bool` would probably be easier here, and is more common.

You can also avoid looping over all of these twice by doing something like this:
```
// Remove duplicates
var deduped []string
seen := make(map[string]bool)
for _, tag := range tags {
  if !seen[tag] {
    seen[tag] = true
    deduped = append(deduped, tag)
  }
}
```

File: pkg/trace/stats/concentrator.go [hunk 33]
```diff
@@ -70,6 +88,9 @@ func NewConcentrator(conf *config.AgentConfig, out chan *pb.StatsPayload, now ti
 		peerSvcAggregation:     conf.PeerServiceAggregation,
 		computeStatsBySpanKind: conf.ComputeStatsBySpanKind,
 	}
+	if conf.PeerServiceAggregation && len(conf.PeerTags) != 0 {
```
Comment (katiehockman) [inline pkg/trace/stats/concentrator.go:91]: You already check the length of `conf.PeerTags` in `prepareTagKeys`, so no need to do it twice.

File: pkg/trace/stats/statsraw.go [hunk 34]
```diff
@@ -60,6 +61,7 @@ func (s *groupedStats) export(a Aggregation) (*pb.ClientGroupedStats, error) {
 	if err != nil {
 		return &pb.ClientGroupedStats{}, err
 	}
+
```
Comment (katiehockman) [inline pkg/trace/stats/statsraw.go:64]:
```suggestion
```

File: releasenotes/notes/add-peer-tags-configuration-79529a2af74e655a.yaml [hunk 35]
```diff
@@ -0,0 +1,12 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    Adds `peer_tags` configuration to allow for more tags in APM stats that
```
Comment (katiehockman) [inline releasenotes/notes/add-peer-tags-configuration-79529a2af74e655a.yaml:11]: Please add more details here. For example, how should it be configured? You can use this as an example: https://github.com/DataDog/datadog-agent/blob/main/releasenotes/notes/apm-memcached-keep-command-aecd5105a09e1235.yaml

**Review by katiehockman — COMMENTED** [review]
I didn't look closely over the tests, but I'm hoping the other reviewers who have more context about the changes have done this and verified that they look correct.

File: pkg/config/config_template.yaml (hunk 22 above)
Comment (jdgumz) [reply pkg/config/config_template.yaml:1488]: Right! Will fix.

File: pkg/trace/stats/concentrator.go (hunk 32 above)
Comment (jdgumz) [reply pkg/trace/stats/concentrator.go:59]: Interesting, I learned that `map[string]struct{}` was more idiomatic for this thing a while back?
I'm not opposed to using bool though.

The suggestion to dedupe in the same first pass is good, will do.

File: pkg/trace/stats/concentrator.go (hunk 31 above)
Comment (pkalmakis) [reply pkg/trace/stats/concentrator.go:67]: I recommended sorting in the other function to defend against future programmers using the hash function incorrectly (without sorting the tags first). It would be a slight optimization to remove it, but we've mitigated the additional sort with first checking if the list is already sorted.

File: pkg/trace/stats/concentrator.go (hunk 32 above)
Comment (pkalmakis) [reply pkg/trace/stats/concentrator.go:59]: In datadog's own backend, it's much more common to use a `struct{}` as it uses less space I believe. I like the idea of modifying both datastructures in a single loop.

File: pkg/trace/stats/aggregation.go (hunk 29 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:150]: I think you're right. This is the first part of the client stats path where the incoming data _may_ have peer.service and `PeerTags` set. If they're nil, then the hash should be 0 as expected.

During the aggregation we need to check if we're enabled because that determines whether we should include or drop that data (and re-aggregate accordingly). In the end I don't think we have to check if we're enabled here. Let me know if that makes sense.

File: pkg/config/config_template.yaml (hunk 21 above)
Comment (jdgumz) [reply pkg/config/config_template.yaml:1485]: Discussed offline; public docs are coming and I will mark `[BETA]` in the description.

File: pkg/proto/datadog/trace/stats.proto (hunk 23 above)
Comment (jdgumz) [reply pkg/proto/datadog/trace/stats.proto:75]: +1, will revise

File: pkg/trace/stats/aggregation.go [hunk 36]
```diff
@@ -80,15 +88,47 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+	var pt []string
+	for _, t := range peerTagKeys {
+		if v, ok := s.Meta[t]; ok {
+			pt = append(pt, t+":"+v)
+		}
+	}
+	return pt
+}
+
+const hashDelimiter = " "
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:115]: This is not a safe delimiter as spaces can be in the tags

File: pkg/trace/stats/concentrator.go (hunk 31 above)
Comment (katiehockman) [reply pkg/trace/stats/concentrator.go:67]: +1 sounds good. Either way, let's not do it twice 😄

File: pkg/trace/stats/aggregation.go (hunk 28 above)
Comment (jdgumz) [reply pkg/trace/stats/aggregation.go:128]: Discussed offline; the use of `strings.Join` results in 2 extra allocs.

**Review by katiehockman — APPROVED** [review]
Approval (upon fixing the hashing issue we discussed offline, since it sounds like `strings.Join` could have unanticipated performance costs)

**Review by katiehockman — APPROVED** [review]

File: pkg/trace/stats/aggregation.go [hunk 37]
```diff
@@ -80,23 +88,61 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+	}
+	if !sort.StringsAreSorted(tags) {
+		sort.Strings(tags)
+	}
+	h := fnv.New64a()
+	for i, t := range tags {
+		h.Write([]byte(t))
+		if i != len(tags)-1 {
+			h.Write([]byte{0})
+		}
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:127]:
```suggestion
		if i > 0 {
			h.Write([]byte{0})
		}
		h.Write([]byte(t))
```
slightly simpler alternative

File: pkg/trace/stats/aggregation.go [hunk 38]
```diff
@@ -80,23 +88,61 @@ func NewAggregationFromSpan(s *pb.Span, origin string, aggKey PayloadAggregation
...
+	}
+	return h.Sum64()
 }
 
 // NewAggregationFromGroup gets the Aggregation key of grouped stats.
 func NewAggregationFromGroup(g *pb.ClientGroupedStats) Aggregation {
-	return Aggregation{
+	agg := Aggregation{
```
Comment (pkalmakis) [inline pkg/trace/stats/aggregation.go:134]:
```suggestion
	return Aggregation{
```

File: pkg/trace/stats/client_stats_aggregator.go [hunk 39]
```diff
@@ -259,6 +262,7 @@ func (b *bucket) aggregationToPayloads() []*pb.ClientStatsPayload {
 				HTTPStatusCode: aggrKey.StatusCode,
 				Type:           aggrKey.Type,
 				Synthetics:     aggrKey.Synthetics,
+				PeerTags:       counts.peerTags,
```
Comment (pkalmakis) [inline pkg/trace/stats/client_stats_aggregator.go:265]: We'll need a hash on the `aggrKey` to make this work. I think i'm beginning to see why you wanted to leave this out...

**Review by pkalmakis — APPROVED** [review]

File: pkg/trace/stats/concentrator.go (hunk 31 above)
Comment (jdgumz) [reply pkg/trace/stats/concentrator.go:67]: Agreed - we should ultimately be sorting once. In the hash func we'll check to see we're sorted, which will be an extra O(n) cost, but we at least won't sort twice.
